### ISSUE TITLE
Would be fun to experiment with supersaw spread algorithms

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#242]
One thing I stumbled in my experiments years ago are a few different ways of spreading oscillator pitches.

1. prime number frequency so there is the least possible beating (not actually sure how to implement fractional prime number spread)
 * oh I wonder if it's just a matter of assigning prime numbers like, say we have 4 oscillators. Assign 2, 3, 5, 7. Now multiply by a spread factor:  .01 spread = .02 .03 .05 .07 hz offset... hmm, and then maybe you could center it. -0.025 -.015 +.005 +.025

2. Apply noise with different filtering amount (slow vs fast noise), and spread evenly.

3. Spread evenly

OH! This reminds me. I need to create a synth that implements my orchestral supersaw:

http://www.elanhickler.com/_/saw%20portamento%20expiriment.mp3

**Comment by elanhickler** [comment]
I found the old thread that describes using primes to create the best frequency spread: https://www.kvraudio.com/forum/viewtopic.php?p=5444677#p5444677

**Comment by RobinSchmidt** [comment]
hmm interesting. more specifically, it uses differences of square-roots of primes (in the stuff, aciddose posted). i think, primes numbers themselves are probably not as useful as irrational numbers. one could also consider reducing the amplitudes of the "outer" frequencies by some rule

**Comment by elanhickler** [comment]
How about another idea, a supersaw oscillator, this would not allow my saw orchestra idea, but it will allow for more controlled spread. Each point represents a saw pulse, and you move the points left and right randomly, and they should never cross

here's my poor attempt to draw it:
![image](https://user-images.githubusercontent.com/9116575/46262021-496d7b00-c4b0-11e8-8d09-b581fe6aa9d8.png)

I think this is how the Yusynth Saw Animator works: http://yusynth.net/Modular/EN/SAWANIM/index.html

**Comment by RobinSchmidt** [comment]
hmm...this is probably quite messy to implement and doesn't straightforwardly generalize to other waveforms. i want also supersquares, supertriangles, etc. :-)

**Comment by elanhickler** [comment]
Ok let's make some progress on this.

What oscillator algorithm do you suggest I use from your library for low CPU antialiased sawtooth? Up to 32 saws and 8 voices is ideal.

**Comment by RobinSchmidt** [comment]
maybe try rosic::SuperOscillator. if i remember correctly, this is the class that also jamal used for his synth...hmm...or - well - you just want sawtooth and don't really care about other waveforms?

**Comment by elanhickler** [comment]
correct, other waveforms are inferior and should be exterminated.

**Comment by RobinSchmidt** [comment]
hmm - i actually tend to like supersquares, too. maybe this could be a good opportunity to put my new blep code to use. i have this new rsBlepReadyOsc class in the rs_testing module. it does not by itself produce an anti-aliased wave, though - it would need to be embedded into some sort of driver object together with the blep

**Comment by elanhickler** [comment]
yeah ill add the other shapes if you make them available

**Comment by RobinSchmidt** [comment]
i'm just creating a "driver" class for a dual saw/pulse oscillator that will use bleps for the waveform discontinuities and hardsync. this will be a good showcase, how the whole blep-system is supposed to be used

**Comment by elanhickler** [comment]
p.s. I can't do anything with your new code until you upgrade your JUCE, my latest personal library for audio plugins depends on it. I've halted Soundemote progress until you upgrade.

**Comment by elanhickler** [comment]
Hey Robin, I've been thinking I might want to hire someone to work on Soundemote plugins, but then I was thinking why not hire you? Maybe this will motivate you to upgrade your JUCE. Get snowflake more or less working, I can get the UI done, you get 50% of profit, and then I have more projects you can create for Soundemote. Then we do all the marketing and you continue to do DSP, win win.

**Comment by RobinSchmidt** [comment]
i just had a discussion on supersaws yesterday and was told that the originial actually *did* alias a lot and that it was therefore common to just use a highpass belowthe fundamental to get rid of the worst (sub-fundamental) aliasing components. that gave me the following idea:



- let each saw-stack run at its own sample-rate which is selected to be an exact  multiple of the desired fundamental frequency
- the saws themselves are then generated naively
- aliased components will (hopefully) be a pleasant enrichment because they will be  harmonically related to the fundamental
- adding more saws to the stack will scale well, like 
   total cost = N * (cost of naive saw) + resampling-cost
 instead of:
   total cost = N * (cost of anti-aliased saw)
 if N is large, we get a payoff approaching N * (cost of anti-aliasing)
- maybe integer/fixed point arithmetic should be used - and SIMD

...we would need an on-the-fly resampler (per voice)...something that may be a generally useful addition to the library anyway. i'd actually like to have an oscillator somewhat more flexible - kinda like the TriSaw but also with convenient access to all sorts of square/pulse-waves ...so maybe more like a PulseSaw rather than TriSaw - but it should also be cheap enough, to stack a big bunch of them without frying the CPU. trisaw has a division per sample (for the shape) - which is expensive and i'd like to avoid that. ...i have some ideas already...about piecing together a waveform from linear segments

but when thinking about it...the most flexible and cheap way to generate all sorts of waveforms is probably still the good old wavetable osc. ...of course, with the disadvantage, that certain waveform features can't easily be modulated in realtime - it has all its pros and cons. but if all you care about is the fat stack of oscs, then it's probably hard to beat in terms of bang-for-the-buck a

**Comment by RobinSchmidt** [comment]
>  I've been thinking I might want to hire someone to work on Soundemote plugins, but then I was thinking why not hire you?

 
 hmm - how would that change anything in the way, i work for you? you mean, i should work more directly in your codebase? for snowflake - yes - i will finish the DSP part, you can slap a nice gui on it and we'll make it 50/50. i will also create a lot of presets, the user manual and some video tutorials. in fact, i actually want to create some more educational video content in the future about all sorts of sound synthesis - including, but not limited to, fractal pattern synthesis (as i have dubbed the stuff, that snowflake does). ..but also more standard stuff like subtractive, FM, additive, etc. - probably using liberty as much as possible. i have agreed to take over some sort of teaching position at a private audio school starting some time in summer - and i will have to start preparing lecture material soon. but i still want to continue to work on dsp code and i actually think, this could all fit together quite well and there should be synergies

**Comment by elanhickler** [comment]
Yes, work directly on my codebase + I have some money I could pay you hourly for specific things, and other things you'd be paid in royalties. I need you to fix some crashes and develop your UI library a little bit. Also, I want help implementing the "click to add module" type interface.

**Comment by RobinSchmidt** [comment]
soo - that would basically more or less boil down to switch from a by-the-hour payment to partially-by-the-hour/partially-royalty-based but otherwise not much would change? 

damn! that sync stuff is more tricky than expected - i'm almost there:
www.rs-met.com/sounds/generators/SyncPhasorNaive.wav
www.rs-met.com/sounds/generators/SyncPhasorMinBlep.wav
...but there are still artifacts - i think, i'm not yet handling double-resets correctly (when master and slave have both a reset within one sample)...but the results sound promising - aliasing seems to be cleaned up quite well - if i can only get rid of these artifacts

**Comment by elanhickler** [comment]
yes

sync sounds good!

**Comment by elanhickler** [comment]
actually id rather you be paid 100% in royalties, i hope soundemote in the future will generate that much revenue that you essentially make more than you would with an hourly rate

**Comment by RobinSchmidt** [comment]
hmm...well - that could work as a long term perspective but i would need some sort of gradual transition because i currently really depend on a somewhat steady by-the-hour payment. but for the more creative r&d heavy sort of work, i actually already sometimes didn't write work-hours, because it's somehow really undesirable to do such things while a clock is ticking. i want to account fairly and only charge for productive time and it sometimes just takes a lot of time to read up on things, experiment, draw scribbles, do math or just sit there and *think* and it feels somehow not like productive work time. 

for other types of work (less creative, more grunt-work or just follow-a-plan), it's more reasonably possible to correctly account for the work. these would be things updating juce or following through with implementing a design plan that already has been laid out. but the figuring-stuff-out time is hard to account for. for example, i didn't charge for the r&d on the envelope-follower refinements because most of the time was spent on figuring out *how* to do it. i could have been keeping track of the actual programming time ...but then didn't because it was so much mixed up with my learning time

**Comment by elanhickler** [comment]
Yes, hourly for now until royalties can meet and exceed hourly pay.

**Comment by RobinSchmidt** [comment]
i think, i got the sync working:
www.rs-met.com/sounds/generators/SyncSweepNaive.wav
www.rs-met.com/sounds/generators/SyncSweepPolyBlep1.wav
www.rs-met.com/sounds/generators/SyncSweepPolyBlep2.wav
www.rs-met.com/sounds/generators/SyncSweepMinBlep.wav
www.rs-met.com/sounds/generators/SyncSweepOversampling.wav
well...at least, the blep versions sound similar to the oversampled (naive) version - so unless i messed up the oversampled naive version too, the results seem ok. i wonder a bit about LFO'ish wobble though - if i should expect that

**Comment by elanhickler** [comment]
Check this out, we can isolate the antialiased junk almost perfectly by doing Naive minus PolyBlep1

http://www.elanhickler.com/transfer/SyncSweepNaiveSubtractPolyBlep1.wav

Blep1 is most dark in terms of lacking in high frequency but is closest to the real world shape of the oscillator
Blep2 is not as dark, a compromise between best shape and best sound
Minblep sounds closest to naive oversampling and has the most high frequency but has the worst shape
![ezgif-5-1993e5d7f4da 1](https://user-images.githubusercontent.com/9116575/57157517-c2f68000-6d95-11e9-8ba5-3ee615b76b3b.gif)

**Comment by RobinSchmidt** [comment]
> we can isolate the antialiased junk almost perfectly by doing Naive minus PolyBlep1. Using PolyBlep2 for this does not isolate the antialiasing as well.

yay! pure ugly aliasing! did you compensate the delay? PolyBlep1 has a delay of 1 sample and PolyBlep2 a delay of 2 samples

**Comment by elanhickler** [comment]
yes I compensated for delay

**Comment by elanhickler** [comment]
Looks like you are writing some code for supersaw frequency spread.

Did you catch that the new Valhalla delay uses the irrational golden ratio to spread delay tap timings for best sound? That could be used for sawtooth frequency somehow. Not sure how it works.

**Comment by elanhickler** [comment]
can't wait to hear your supersaw spread demos

**Comment by elanhickler** [comment]
I made a thread for our joint project subtractive synth: https://www.kvraudio.com/forum/viewtopic.php?f=1&t=524814&p=7392668#p7392668

**Comment by RobinSchmidt** [comment]
> Looks like you are writing some code for supersaw frequency spread. ... can't wait to hear your supersaw spread demos

hmm...very very preliminary pre-preview:

www.rs-met.com/sounds/generators/SuperSawMetallicFracPart.wav

the "metallic" in the file name refers to the fact that i'm using (the fractional parts of) metallic ratios:

https://en.wikipedia.org/wiki/Metallic_mean

sounds already okayish to my ears - but the higher pitched notes somehow suffer from some sort of "flutter" - it is as if i can hear the peak-envelope or something which sounds a bit like a low freq hum or rumble. tried the old trick of the highpass at the fundamental - doesn't really help much ...still very early stages

> Did you catch that the new Valhalla delay [maybe] uses the irrational golden ratio to spread delay tap timings

no - is there some info about that somewhere? i'm currently think about an algo to spread the frequencies based on recursive subdivision of an interval: 
- take the original freq interval 
- split it at a ratio of r
- choose the largest in the current set of intervals and split it again at the same ratio
- ...and so on

> I made a thread for our joint project subtractive synth

aha! nice! i will keep track of it!

**Comment by RobinSchmidt** [comment]
...but the bleps work really well for this - even the 1-sample delay compensation gives good results where without anti-aliasing, the results are an ugly mess. and the best is: we can use take a single blep object for the whole stack of saws - we don't need a separate blep for each saw.

**Comment by elanhickler** [comment]
I don't know of any specific info on using the golden ratio for time or frequency.

I'll experiment with noise-based frequency spread when I get to creating the synth.

**Comment by elanhickler** [comment]
Oh, another thing to improve the sound of supersaw is to make the high frequencies spread less and the low frequency spread more. I don't know if linear or exponential curve over frequency would work here.

**Comment by RobinSchmidt** [comment]
> another thing to improve the sound of supersaw is to make the high frequencies spread less and the low frequency spread more

i think, we may leave that to the sound designer, who can just use the key as mod-source for the detune parameter. if it's common enough to do, we may also add detune keytracking directly to the osc class - we'll see

>  I need to hear it against other examples to hear what works more or less

i'll add a module to ToolChain for this osc soon, so we can select the different spreading strategies to experiment with there

**Comment by elanhickler** [comment]
no sound from the new BlepOsc 😭

**Comment by RobinSchmidt** [comment]
> no sound from the new BlepOsc sob

it's still buggy as hell...with Density=1, i get infs - set it to some other value. the prime-diff distributions still cause access violations....and the detune behaves wrong. ...probably not yet worth to check out

**Comment by elanhickler** [comment]
that's a tough choice. I say go for supersaw and maybe I can post some audio demos for our fans over at KVR.

**Comment by RobinSchmidt** [comment]
ok - i't now halfway ready to for some first noodling. currently, the saws start all in phase, that's why there are transients. i want to introduce an initial start-phase spread parameter by which this transient effect can be controlled. later, i also want to make the density parameter continuous by by gradually fading in one saw after another when ramping up density.

...but for now i must leave for the superbooth. will play around with it tomorrow....

**Comment by elanhickler** [comment]
I'm getting some really interesting sounds with all the algorithms.

oH you forgot panning! Half saws should be panned hard left, half panned hard right.

**Comment by elanhickler** [comment]
OK! I love all the algorithms. I want all of them for the final product. The algorithm used for spread can be part of the advanced controls so as not to clutter the interface. The default algorithm will be Prime Square Root, that is the best algorithm for general purpose supersaw. But if you were planning on adding more algorithms, do it! Each one is great.

So when we start building the instrument I need access to the individual sawtooths so I can do things like:
* modulate the frequency of each sawtooth with an individual noise generator for random frequency spread
* do my orchestral spread stuff with individual portamento per saw
* different panning spreads... maybe even amplitude spreads

**Comment by elanhickler** [comment]
posted a demo: https://www.kvraudio.com/forum/viewtopic.php?p=7395723#p7395723

**Comment by elanhickler** [comment]
Range Split Odd and Even seem sound the same

Metallic, Range Split Odd, Range Split Even, Range Split Skewed all seem to sound the same if DistParam is 0.5.

**Comment by RobinSchmidt** [comment]
i added the parameter for the initial phase coherence (hmm...we could flip it left/right and call it initial phase spread...which one is better? i tend to think, the current way). when it's set to zero (phases maximally spread out at start), there are still some transients - but they sound different now. they have a flangy character now as opposed to the sort of thud, when they start coherently. and the prime sqrt *diff * has the most neutral transient - actually next to none - so i would vote to take that as default.

> Range Split Odd and Even seem to sound the same

yep - they have the same quality - but i just checked the numbers (if i may have abug) - but they are different indeed. maybe we can experiment with crossfading/morphing between them. it's conceivable that these lead qualitatively same results at both ends and something different happens in the middle. we'll see. unfortunately, the algorithms are expensive (they scale with O(N^2)) and i'm not sure, if that can be improved

i didn't actually forget about panning - it's just not yet implemented. and yes - i want to introduce a amplitude distribution parameter to apply a sort of bell curve to the clusters of sines at the harmonics. i actually have some ideas how to introduce a (chaotic feedback) modulation to the frequencies of the saws (mainly to get rid of the remaining regularities)

**Comment by RobinSchmidt** [comment]
a lot of fun indeed. i added a handful of presets to my preset repo. the sonic range of this oscillator array goes way beyond supersaw. ...imagine what it will do when we have morphable waveshapes! i want some sort of continuous morph between saw-up/saw-down/pulse/triangle etc. ...similar to the trisaw osc but with out the bending function (because that's expensive and thwarts the blep scheme). just a waveform pieced together from linear segments - with all their steps and corners blepped and blamped

**Comment by elanhickler** [comment]
Can you add some bad or silly spread algorithms? Like basic linear, basic exponential, basic logarythmic, whatever you think would make a bad or undesirable spread.

**Comment by elanhickler** [comment]
posted another demo:

**Comment by elanhickler** [comment]
posted another demo: https://www.kvraudio.com/forum/viewtopic.php?p=7396871#p7396871

btw, I went ahead and added stereo myself for the last two demos in code.

**Comment by RobinSchmidt** [comment]
> Can you add some bad or silly spread algorithms?

ok. i had considered this myself already - but with some sort of continuous adjustment between silly and not-silly....

> I went ahead and added stereo myself for the last two demos in code.

really? where is this code? how exactly did you implement this? 

> your Prime Power algorithm probably beats every product on the market

really? is this better than prime sqrt? i mean - of course it is more flexible, since in contains prime-sqrt as special case for 0.5. i was just trying to make some reasonable use of the adjustable parameter (which was otherwise dormant for the earlier prime sqrt - which felt weird, user experience wise) - so i thought: ok, let's generalize from prime^0.5 to prime^p for some adjustable power p.

**Comment by elanhickler** [comment]
To add stereo I just added T* L, T* R to the "getSample" methods, and at the base when you go through the oscs I have `i % 2 = 0 ? L+=out : R+=out`

I went with lowest distpower. 0 causes crash. Difficult to hear the effect of this parameter but I know 1 wasn't sounding as good.
![image](https://user-images.githubusercontent.com/9116575/57576346-540ecc00-7412-11e9-907a-8a87a95f4f1a.png)

**Comment by elanhickler** [comment]
There are two noise-based frequency spread algorithms I know of. You don't have to add these features. I will if you don't in the final product:

1. Change the frequency of the sawtooth randomly for every cycle with a control for the random range, which is 0 to 1. This means high notes are more wild in frequency and low notes are cleaner. The equation is this: `frequency + frequency*randomValue*range`.  This is how Tone2 Saurus works, but I actually discovered this method before knowing Saurus implemented it.

2. Change frequency continuously with a noise signal where we have two controls, one for the noise shape (slow or fast changes, maybe that is a lowpass filter cutoff on the noise), and another control for range. This is how Audjoo Helix works and other synthesizers work.

**Comment by RobinSchmidt** [comment]
> To add stereo I just added T* L, T* R to the "getSample" methods, and at the base when you go through the oscs I have i % 2 = 0 ? L+=out : R+=out

but how do you deal with the bleps? to produce an anti-aliased stereo signal, we need two blep objects - one for each channel - and must multiply the passed step amplitudes with the gain values of the respective channel. i actually already have a comment about this at the bottom of OscArrays.h:
```cpp
  // todo: have an array of pan positions - if we do stereo later, we will need two blep objects 
  // - one for each channel
};
```
if you didn't do it  this way, i'd expect the aliasing to come back. ...but don't bother  - i'll implement all this myself. i know exactly, how it has to be done.

> I went with lowest distpower. 0 causes crash. 

oh - yes - i already encountered this myself. it hits a numeric singularity (like div-by-zero) or something. i'll see, how to handle this...maybe catch it as special case...

> There are two noise-based frequency spread algorithms I know of....

i actually had a talk yesterday with niko herz at superbooth, where he explained to me in some detail, how helix does it - but i probably shouldn't reveal these things here in this semi-public place - dunno. he did some sort of consulting with the developer of helix about these things. yes, it is noise modulation of the individual frequencies of the saws - with a special kind of noise. ...however, i would expect that the exact kind of noise is perhaps not that critical anyway. my idea is: use the output of the supersaw, isolate the cluster around the fundamental with a bandpass, modulate it down to 0 Hz (ringmod + lowpass) - use that derived "lowpass-noise" signal for modulating the frequencies. due to the feedback nature of the algo, i'd expect it to produce chaotic behavior. if we just linearly modulate all the frequencies with the same signal, we actually change their frequency ratios (f1+d)/(f2+d) != f1/f2 - but we probably should apply the mod signal alternatingly (+/-) to successive saws. ...we'll see, ho this works out....

**Comment by RobinSchmidt** [comment]
i added parameters for randomizing the initial phases (amount and seed). the start phases determine (obviously), how the sound starts, i.e. the character of the initial "transient" (i use quotes because the term does not always fit, but you get it). i added a couple of new presets - the different variants of "SynthString" demonstrate that effect nicely (although they also have different detune settings). sound-design wise, i think, it's a good idea to set detune to zero when tweaking the start-phase related parameters until one has a nice initial character - for example, for string sounds without noticable attack, the waveform should initially sound thin. when this is done, dial in the desired detune

**Comment by RobinSchmidt** [comment]
oh - and i fixed a bug  - the increment array was not sorted - now it is - but this may break some presets - but with the new start-phase randomization parameters, it should be possible to arrive at similar sounds again

**Comment by RobinSchmidt** [comment]
working on the stereo spread. just alternatingly panning left and right somehow doesn't seem good enough. we should at least start in the middle and treat even and odd densities differently - because with full spread and an odd density, the number of saws left and right is different - i think, with odd densities, the middle one should be panned to the center or something - i need to experiment a bit

**Comment by RobinSchmidt** [comment]
stereo spread works now. sounds really gooood!

**Comment by elanhickler** [comment]
Just keep in mind I've done extensive study of the Supersaw, the best sound is pan each saw hard left hard right. Anything else is just different flavors and alternatives, which won't sound a good but in the right context it might sound interesting. Just like your amp spread ideas, the best is all saws same volume. Alternative amp schemes will be interesting but not better.

**Comment by elanhickler** [comment]
The next thing to add is the two noise spreading algorithms. If you're done with adding features I can take over. If not:


1. Add noise/frequency spreading features (most important)
2. Add portamento speed and spread (most important)
3. Add different oscillator shapes (could be added as an update)
4. Add morph shape controls (could be added as an update)

After #1 and #2, everything else is flavoring and extras.

What about polyphony?

Oh, I was actually hoping to add a scheme of "distribute sawtooths to each key based on number of keys played", so it would be like... if you have 32 saws enabled, if you play 4 keys at once, you distribute 8 saws per key. I already started on the logic a while back.

**Comment by RobinSchmidt** [comment]
i actually wanted to look into osc-shape next. portamento? as in note-slide? i think, this is the job of some sort of outlying driver object...which may also handle polyphony. polyphony is a rather big and complicated thing.  what might be also interesting is to invert the polarity some of the saws (maybe the even indexed are saw-up, and the odds saw-down)

**Comment by elanhickler** [comment]
Keep in mind I can't sell a monophonic subtractive synth. I could do a simple synth and release for free and have a paid version later. That is also beneficial to gaining an audience.

**Comment by RobinSchmidt** [comment]
making a mono version for free and the poly version paid sounds like a good strategy anyway

**Comment by RobinSchmidt** [comment]
>  the best sound is pan each saw hard left hard right. 

this is what happens if the StereoSpread parameter is +1 or -1. these settings create the maximum stereo width. if i'm not mistaken, it should lead to a (time averaged) correlation of zero between left and right channel (which seems to be indeed the case, looking at a correlation meter). in some way, it's the widest width that is still mono-compatible. but the correlation may temporarily become negative, which may lead to temporary mono-incompatibility issues - so it's good to allow less width. i like settings around +-0.5, too

>  the best is all saws same volume. 

hmm - ok - i didn't try this yet. however - i think, it's rather simple to implement amplitude curves (we have the amplitude multiplier arrays available anyway (because they are needed for the stereo spread) - so amplitude curves is just a matter of applying another formula in their computation). another idea: instead of just applying a sort of bell-curve (making the outer saws more quiet), we could also use comb-like curves - that would lead to some sort of morph between supersaw (~7 saws) and synth-string (>20 saws) - it would attenuate in the *intermediate* saws, not the *outer* ones

**Comment by elanhickler** [comment]
Another practical idea to increase perceived frequency and stereo spread without increasing dissonance would be to have each consecutive saw detuned more and the more detuned the saw, set the amplitude lower. Then you have a strong in-tune signal with a quiet out-of-tune signal which will sound big and wide.

**Comment by RobinSchmidt** [comment]
>  The more detuned the saw, set the amplitude lower. Then you have a strong in-tune signal with a quiet out-of-tune signal 

how is this different from my amplitude bell-curve idea?

**Comment by elanhickler** [comment]
not different in terms of amplitude, just adding that the frequencies of quieter saws can be exaggerated in terms of detune.

**Comment by RobinSchmidt** [comment]
ok - yes, i see. indeed, if we apply amplitudes as function of detune, we may increase the total detune. maybe there should be some kind of coupling between amp-env amount (env in the sense of spectral envelope) and total detune

**Comment by elanhickler** [comment]
Oh, what about using your spread algorithm prime power for delay time spread for lush reverb?

**Comment by RobinSchmidt** [comment]
in principle, the same considerations apply to to the choice of unison frequencies and reverb delay times. so, yes, it could be useful. i have an FDN implementation in the library which we can use as starting point to experiment with this stuff.

**Comment by elanhickler** [comment]
Add a filter and add a smooth AD envelope for amplitude and a seperate one for filter cutoff to our supsersaw WIP synth. I'm excited to experiment with this envelope.

**Comment by RobinSchmidt** [comment]
i'll add the new envelope to toolchain. first monophonically and from there, i'll try to extend it (and with it, the whole modulation system) to the polyphonic case. then i'll need to make the osc polyphonic, too and then my ladder filter. with these 3 components, we can throw together a basic subtractive synth within toolchain itself. for the "generator" types of modules (oscillators, etc.), i plan to just add their outputs - so we can even simulate a multi-osc synth

**Comment by elanhickler** [comment]
exciting!

**Comment by RobinSchmidt** [comment]
i made a simple example patch for toolchain - it uses the trisaw osc instead of the osc array (because the osc array does not yet have an amplitude parameter - but that's a trivial addition) but demonstrates the principle:
```xml
<?xml version="1.0" encoding="UTF-8"?>

<ToolChain PatchFormat="1" ActiveSlot="1">
  <Slot Type="AttackDecayEnvelope">
    <Slot1-AttackDecayEnvelope PatchFormat="1" Attack="2.00049" Decay="219.234"/>
  </Slot>
  <Slot Type="TriSawOscillator">
    <Slot2-TriSawOscillator PatchFormat="1" Amplitude="0"/>
  </Slot>
  <Slot Type="None">
    <Slot3-None PatchFormat="1"/>
  </Slot>
  <Modulations>
    <Connection Source="Slot1-AttackDecayEnvelope" Target="Slot2-TriSawOscillator.Amplitude"
                Depth="1.0" DepthMin="-1.0" DepthMax="1.0" Mode="Absolute"/>
  </Modulations>
</ToolChain>
```
the new envelope is routed to the amplitude of the osc (additively, absolutely - in practice, we'll probably want to choose the multiplicative mode for amp envs - but whatever...this is just a quick demo patch)

**Comment by elanhickler** [comment]
add copy and paste buttons

```cpp
MenuHelper::addCallbackItem(menu, "Copy preset", [&]()
      {
        XmlElement* xmlState = moduleToEdit->getStateAsXml("", true);
        if (xmlState == nullptr)
          return;

        juce::SystemClipboard::copyTextToClipboard(xmlState->createDocument(String()));
      });

    MenuHelper::addCallbackItem(menu, "Paste preset", [&]()
      {
        String clip = juce::SystemClipboard::getTextFromClipboard();
        XmlDocument doc(clip);

        auto state = ScopedPointer<XmlElement>(doc.getDocumentElement());

        if (state == nullptr)
          return;

        setPresetName("from clipboard");
        moduleToEdit->setStateFromXml(*state.get(), "from clipboard", true);

        presetWasChangedFunction();
      });
```

**Comment by RobinSchmidt** [comment]
i actually don't want more buttons there for space reasons, but i guess, such things could be put into a right-click popup menu

**Comment by elanhickler** [comment]
https://hearthis.at/soundemote/adenvelopewithfeedback/

ad envelope with feedback is a little buggy, the effective modulation range is around -6 to 0 in exponential mode. The envelope is a little buggy without feedback as well, the attack/decay times will run into each other if attack is long and decay is short and cause the envelope to go flat.

I start with no feedback then introduce feedback and adjust the decay time and decay modulation going to the filter. Great envelope.

**Comment by RobinSchmidt** [comment]
yes - indeed - the attack time *must* be strictly shorter than the decay time. the formulas require that. i have an rsAssert for that somewhere but don't know yet, how to deal with it. ...probably the most reasonable thing is to clip the attack time to 0.99*decayTime or something. oh - and attack of absolute zero also doesn't work - maybe i can fix that - but maybe not - if not, we must live with a very small (subsample) minimum attack time

**Comment by elanhickler** [comment]
The attack is fine, it doesn't need to be faster.
